scanner/decisions/0012-reindexing.md

```diff
@@ -0,0 +1,192 @@
+    if err != nil {
+        log.Fatal(err)
+    }
+    vscnrs := indexer.MergeVS(pscnrs, dscnrs, rscnrs, fscnrs)
+
+	// Check if the manifest was indexed with the latest versioned scanners.
+    ok, err := i.Store.ManifestScanned(context.Background(), digest, vscnrs)
+    if err != nil {
+        log.Fatal(err)
+    }
+    if ok {
+        // Image already indexed with latest versioned scanners.
+    } else {
+        // Image was either not previously indexed, or it was with older versioned scanners.
+    }
+}
+```
+
+This method does not require any kind of migration to account for previously indexed images, as the data/tables already have all required the information.
+
+#### Alternative - Tracking Indexer State
+
+Scanner V4 Indexer would maintain a table which tracks each successful Index Report (identified by the report's hash ID) 
+alongside the Scanner V4 Indexer's state (equal to the `etag` sent by Clair).
+
+This way, when clients reach out to Scanner V4 Indexer, it may first check not only if the image was already indexed,
+but also the state of the Indexer at the time of indexing. If they do not match, then we will create the full `*claircore.Manifest`
+and re-index the image. As mentioned in the [Context](#context) section, ClairCore already has some builtin performance optimizations for this.
+Another benefit to this that `*claircore.Manifest` continues to only be generated, as needed.
+
+However, there are cons which outweigh the pros:
+
+* More data for Scanner V4 to track outside ClairCore
+  * This is not a blocker, but it is worth mentioning
+* Handling preexisting Index Reports
+  * This is the challenging aspect, which would block this
+  * There are a few options:
+    * Write a migration
+      * This involves doing essentially the same thing `ManifestScanned` does to determine the "versioned scanners"
+      * This may not be able to be pure PostgeSQL, as the `etag` is generated in Go, so that is an extra challenge
+    * Populate the table upon Index Report fetching
+      * If the entry does not exist in the table (the case for all Index Reports created prior to this release), then create the Index Report and populate the table
+      * This essentially forces a re-indexing of all previously indexed images, whether it is required or not, which may be costly
+
+### Random Deletions
+
+When it comes to variables outside "versioned scanners" there is no currently known clean way to handle the re-indexing.
+
+Something like creating our own "versioned scanner" to track other types of changes is not a solution:
+
+* Updating the version of the custom "versioned scanner" will only affect that scanner, and not the others which would all be skipped upon re-indexing
+* It is not feasible to track every single change in ClairCore which may affect Index Reports
+
+Instead, we will randomly delete manifests from the ClairCore database. Deleting a manifest from ClairCore
```

Comment:
nit: Seems that these deletes are not so much random, but rather scheduled or TTL based. Random to me implies unpredictable behavior which I don't believe this is.

Reply:
These are kind of random within a range. I guess it is controlled, as the range is controlled, but the final outcome is random in that range. Does that change anything?

Reply:
What is the purpose for making it semi-random? Why not a configurable time period to make it explainable?  Thinking of a scenario 'why did we reach out to the registry for image A but not image B?'.  Is this is away to introduce 'delete jitter' of sorts?

Reply:
Yes, the purpose is to spread the load out as much as possible (within reason (ie not making the random range from now until the end of time)) so we don't overwhelm the registries. My fear is the scenario where StackRox is introduced to a pre-existing environment and we scan many images all at once in the beginning and then each successive re-indexing is all done at the same exact time some X days/minutes/seconds in the future. The hope is to spread it out to ensure both Scanner V4 does not get overwhelmed and the registries do not get overwhelmed (this could happen when a ton of manifests expire just in time for a Central reprocessing to request scans)

Reply:
I see. Believe this already happens today (mass scans) on every restart of scanner-db / rox upgrade FWIW.

Reply:
yep, and I'm hoping we stop doing that with Scanner V4 with the help of the PVC + these random expirations